### PR TITLE
Add docs for query frontend

### LINKED ISSUE #3015: query-frontend: Docs
cc @yeya24

### PR DESCRIPTION
**Pull request by yeya24**
Signed-off-by: Ben Ye <yb532204897@gmail.com>

* [ ] I added CHANGELOG entry for this change.
* [x] Change is not relevant to the end user.

## Changes

Fixes #3015

### CODE REVIEW

File: docs/components/query-frontend.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+  validity: 6h
+```
+
+### Slow Query Log
+
+Query Frontend supports `--query-frontend.log_queries_longer_than` flag to log queries running longer some duration.
+
+## Naming
+
+Naming is hard :) Please check [here](https://github.com/thanos-io/thanos/pull/2434#discussion_r408300683) to see why we chose `query-frontend` as the name.
```
Comment (pracucci) [inline docs/components/query-frontend.md:65]: Should this be part of the component doc?

Comment (bwplotka) [reply docs/components/query-frontend.md:65]: Yes, it's one of components, no?

Comment (pracucci) [reply docs/components/query-frontend.md:65]: Yes. I was wondering if doc should focus on "why" as well (eg. in this case why the name has been picked).

Comment (bwplotka) [reply docs/components/query-frontend.md:65]: Ah, good point. I think this is good enough - no better place I could think of. The reason is that ppl already ask why this name - so we don't want to repeat the work ;p

File: docs/components/query-frontend.md [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+Query Frontend supports a retry mechanism to retry query when HTTP requests are failing. There is a `--query-range.max-retries-per-request` flag to limit the maximum retry times.
+
+### Caching
+
+Query Frontend supports caching query results and reuses them on subsequent queries. If the cached results are incomplete,
+Query Frontend calculates the required subqueries and executes them in parallel on downstream queriers. Query Frontend can optionally align queries with their step parameter to improve the cacheability of the query results.
+
+Currently, only in-memory cache (fifo cache) is supported. An example config:
+
+```yaml
```
Comment (bwplotka) [inline docs/components/query-frontend.md:51]: this can be autogenerated. Can we do this in this PR maybe? (:

Comment (yeya24) [reply docs/components/query-frontend.md:51]: I didn't know we can auto-generate this before. Let me take a look how to do it.

Comment (bwplotka) [reply docs/components/query-frontend.md:51]: yup, looks on object storage page (:

Comment (yeya24) [reply docs/components/query-frontend.md:52]: When investigating this, I found actually some of our embedmd configs are wrong like `[embedmd]: # "../flags/config_index_cache_in_memory.txt yaml"`, this syntax is not supported. So I also changed them in that commit

Comment (bwplotka) [reply docs/components/query-frontend.md:52]: Thanks!

**Pull request merged**